src/agoradatatools/etl/transform/model_details.py

```diff
@@ -0,0 +1,190 @@
+    biomarkers_df = biomarkers_df.rename(
+        columns={"type": "evidence_type", "measurement": "value"}
+    )
+    pathology_df = pathology_df.rename(
+        columns={"type": "evidence_type", "measurement": "value"}
+    )
+
+    # Convert matching controls and aliases from comma-delimited strings to lists
+    model_info_df["matched_controls"] = model_info_df["matched_controls"].apply(
+        lambda x: [item.strip() for item in str(x).split(",")]
+        if (pd.notna(x) and x != "")
+        else []
+    )
+
+    model_info_df["aliases"] = model_info_df["aliases"].apply(
+        lambda x: [item.strip() for item in str(x).split(",")]
+        if (pd.notna(x) and x != "")
+        else []
+    )
+
+    # Build the final data structure
+    result = []
+
+    # Process each model
+    for _, model_row in model_info_df.iterrows():
+        model_name = model_row["model"]
+
+        # Get genetic info for this model
+        genetic_info = []
+        model_alleles = allele_info_df[allele_info_df["model"] == model_name]
+
+        for _, allele_row in model_alleles.iterrows():
+            gene_info = {
+                "modified_gene": allele_row["gene"],
+                "ensembl_id": allele_row["gene_ensembl_id"],
+                "allele": allele_row["allele"],
+                "allele_type": allele_row["allele_type"],
+                "mgi_allele_id": allele_row["mgi_allele_id"],
+            }
+
+            # If it's a human transgene, replace the ensembl_id with the human one
+            if (
+                allele_row["mgi_allele_id"]
+                in human_transgene_allele_map_df["mgi_allele_id"].values
+            ):
+                matching_row = human_transgene_allele_map_df[
+                    human_transgene_allele_map_df["mgi_allele_id"]
+                    == allele_row["mgi_allele_id"]
+                ]
+                if (
+                    len(matching_row) > 0
+                    and matching_row.iloc[0]["gene_symbol"] == allele_row["gene"]
+                ):
+                    gene_info["ensembl_id"] = matching_row.iloc[0]["ensembl_id"]
+
+            genetic_info.append(gene_info)
+
+        # Process biomarkers data
+        model_biomarkers = []
+        model_biomarkers_df = biomarkers_df[biomarkers_df["model"] == model_name]
+
+        # Group biomarkers by evidence_type, tissue, and age_death
+        for (evidence_type, tissue, age), group in model_biomarkers_df.groupby(
+            ["evidence_type", "tissue", "age_death"]
+        ):
+            biomarker_entry = {
+                "model": model_name,
+                "evidence_type": evidence_type,
+                "tissue": tissue,
+                "age": f"{age} months",
+                "units": group["units"].iloc[0],
+                "data": [],
+            }
+
+            # Add individual data points
+            for _, row in group.iterrows():
+                data_point = {
```

Comment:
Another maybe -- If you use `row[["genotype", "sex", "individual_id", "value"]].to_dict(orient="records")`, this should give you a list of dicts with the proper keys and a single value for each key. If that works, you can also apply that to anywhere else applicable.

Reply:
Addendum: I wonder if, by merging some of the data frames at the beginning of this function by `model`, you can kind of... mass do a bunch of this function with `.to_dict()` and eliminate some for loops inside here? Again, haven't played around, just thinking of ways to make the function a little shorter and easier to follow.